src/server.rs

```diff
@@ -58,32 +71,87 @@ pub(crate) async fn start_server(cancellation_token: CancellationToken) {
     let contents = FILE_DESCRIPTOR_SET.to_vec();
     let proto_descriptor_service = tonic_reflection::server::Builder::configure()
         .register_encoded_file_descriptor_set(contents.as_slice())
-        .build()
+        .build_v1()
         .unwrap_or_else(|err| {
             status_err!("failed to build descriptor service: {}", err);
             std::process::exit(1)
         });
     let config = APP.config();
-    let server_config = load_server_config(&config).await.unwrap_or_else(|err| {
-        status_err!("failed to load server config: {}", err);
-        std::process::exit(1)
-    });
-
-    info!("listening on {}", server_config.address);
-    if let Err(err) = tonic::transport::Server::builder()
-        .tls_config(server_config.tls_config.expect("must have tls config"))
-        .unwrap_or_else(|err| {
-            status_err!("failed to configure server for tls: {}", err);
-            std::process::exit(1)
-        })
-        .add_service(ContractCallServiceServer::new(CorkHandler))
+    let server_config = auth_config(&config)
+        .await
+        .expect("failed to get server config");
+    let address = socket_addr(&config).expect("failed to get socket address");
+    let svc = Routes::new(ContractCallServiceServer::new(CorkHandler))
         .add_service(StatusServiceServer::new(StatusHandler))
         .add_service(proto_descriptor_service)
-        .serve_with_shutdown(server_config.address, cancellation_token.cancelled())
+        .prepare();
+    let http = Builder::new(TokioExecutor::new());
+    let listener = TcpListener::bind(address)
         .await
-    {
-        status_err!("server error: {}", err);
-        std::process::exit(1)
+        .expect("failed to bind to address");
+    let tls_acceptor = TlsAcceptor::from(Arc::new(server_config));
+
+    // Start accepting connections
+    info!("listening on {}", address);
+    loop {
+        tokio::select! {
+            // Handle incoming connections
+            incoming = listener.accept() => {
+                let (connection, address) = match incoming {
+                    Ok(incoming) => incoming,
+                    Err(e) => {
+                        eprintln!("Error accepting connection: {}", e);
+                        continue;
+                    }
+                };
+
+                let http = http.clone();
+                let tls_acceptor = tls_acceptor.clone();
+                let svc = svc.clone();
+
+                tokio::spawn(async move {
+                    let mut certificates = Vec::new();
+
+                    let conn = tls_acceptor.accept(connection).await.unwrap();
+
+                    let (_, info) = conn.get_ref();
+
+                    if let Some(certs) = info.peer_certificates() {
+                        debug!("found {} peer certificates", certs.len());
+                        for cert in certs {
+                            certificates.push(cert.clone());
```

Comment:
I have this concern with pushing the peer certificate into a global static because certificates from multiple sessions with different peers connecting get mixed together. 

So like first good peer connects and while good peer is connected evil peer is connected and it treats evil peer as authorized because good peer showed up with a valid certificate. 

I am not 100% sure this will happen in the current code but I don't see anything that really stops this from happening and pattern seems high risk.

Reply:
so this block is a task that gets uniquely spawn each connection, so the certificates vec is owned by a single scope that is unique per client call. is there a global static somewhere I'm missing?

Reply:
I see it now. Okay you are correct.